### PR TITLE
Make `repository` be able to update its client configs.

### PR DESCRIPTION
**Pull request by zhenghuiwang**
This PR makes a `repository` be able to update its own client configurations:

- Change the `ClientConfig` definition, which defines the request for updating a client (also the response for querying a client configs).

- Client update logic in `Repository.UpdateClientConfigs`

- On testing setup: 1. Copy `example-gateway` to a local repository. 2. Update a client. 3. Compare the configuration with hand-written one in `example-gateway`.

r: @uber/zanzibar-team

### CODE REVIEW

File: backend/repository/client_update.go [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+	}
+	return cfgJSON
+}
+
+// UpdateClientConfigs updates JSON configuration files for a client update request.
+func (r *Repository) UpdateClientConfigs(req *ClientConfig, clientCfgDir, thriftFileSha string) error {
+	if err := validateClientUpdateRequest(req); err != nil {
+		return err
+	}
+	cfgJSON := NewClientConfigJSON(*req)
```
Comment (ChuntaoLu) [inline backend/repository/client_update.go]: why not `NewClientConfigJSON` take a pointer as well?

File: backend/repository/client_update.go [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+	if req.Type == "tchannel" && req.Timeout == 0 {
+		req.Timeout = 10000
+	}
+	if req.Type == "tchannel" && req.TimeoutPerAttempt == 0 {
+		req.TimeoutPerAttempt = 10000
+	}
+	return nil
+}
+
+type clientModuleJSONConfig struct {
```
Comment (ChuntaoLu) [inline backend/repository/client_update.go]: overlap with https://github.com/uber/zanzibar/blob/master/codegen/gateway.go#L112, we should establish single SOT to ensure consistency.

File: backend/repository/config.go [hunk 3]
```diff
@@ -78,10 +78,15 @@ type EndptMidConfig struct {
...
 // ClientConfig stores configuration for an client.
 type ClientConfig struct {
-	ID             string
-	Type           ProtocolType
-	ThriftFile     string
-	ExposedMethods map[string]string
+	Name              string            `json:"name"`
+	Type              ProtocolType      `json:"type"`
+	ThriftFile        string            `json:"thriftFile,omitempty"`
+	MuttleyName       string            `json:"muttleyName,omitempty"`
```
Comment (ChuntaoLu) [inline backend/repository/config.go]: This name is not straightforward without certain context.

File: backend/repository/config.go [hunk 4]
```diff
@@ -117,8 +122,13 @@ type ThriftMeta struct {
 type ProtocolType string
 
 const (
-	gatewayConfigFile = "gateway.json"
+	gatewayConfigFile     = "gateway.json"
+	productionCfgJSONPath = "config/production.json"
+	clientConfigFileName  = "client-config.json"
+	clientModuleFileName  = "clients-config.json"
```
Comment (ChuntaoLu) [inline backend/repository/config.go]: `clientsConfigFileName` is arguably a better name from module system point of view, despite it's awfully close to `clientConifgFileName`

File: backend/repository/data/client/baz_client_update.json [hunk 5]
```diff
@@ -0,0 +1,33 @@
+{
+    "_comment_": "JSON for updating a tchannel client.",
```
Comment (ChuntaoLu) [inline backend/repository/data/client/baz_client_update.json]: The JSON schema way is to name this field "description", http://json-schema.org/latest/json-schema-validation.html#rfc.section.7.2

File: backend/repository/client_update.go (hunk 1 above)
Comment (zhenghuiwang) [inline backend/repository/client_update.go]: will change it.

File: backend/repository/client_update.go (hunk 2 above)
Comment (zhenghuiwang) [inline backend/repository/client_update.go]: You're right. ( wired: /codegen/gateway.go@`master`#L112 doesn't have `dependencies`. `dependencies` in the hand-written config seems not used at all!?)

File: backend/repository/config.go (hunk 3 above)
Comment (zhenghuiwang) [inline backend/repository/config.go]: `ServiceName` is overload for http. what could be a better one? `TchannelServiceName`, `TchannelRoutingName`?

File: backend/repository/config.go (hunk 4 above)
Comment (zhenghuiwang) [inline backend/repository/config.go]: that's why i use `clientModule`, meaning for all clients but also distinctive?

File: backend/repository/data/client/baz_client_update.json (hunk 5 above)
Comment (zhenghuiwang) [inline backend/repository/data/client/baz_client_update.json]: Thought about it: JSON schema reserves string "description", which could be a key and is not good for the general purpose of a JSON. `_description_` or `_comment_` `__description__` or `__comment__` all seem good to me (they should be the first one lexicographically).

File: backend/repository/client_update.go (hunk 2 above)
Comment (ChuntaoLu) [inline backend/repository/client_update.go]: Dependencies are took care of at module level, we can add it to the clientConfig struct

File: backend/repository/config.go (hunk 3 above)
Comment (ChuntaoLu) [inline backend/repository/config.go]: Not sure what you mean by overload. How about we use `ServiceName` to denote service discovery/routing name, and `ThriftServiceName` for service defined in thrift file?

File: backend/repository/config.go (hunk 4 above)
Comment (ChuntaoLu) [inline backend/repository/config.go]: maybe `clientsModule`?

File: backend/repository/data/client/baz_client_update.json (hunk 5 above)
Comment (ChuntaoLu) [inline backend/repository/data/client/baz_client_update.json]: makes sense

**Pull request merged**